### ISSUE TITLE
[TODO] A more feature rich vimrc, and a demo gif?

### ISSUE THREAD

**Issue by roxma** [ncm2/ncm2#19]

**Comment by Neur1n** [comment]
In **Python**, I can insert a `class` snippet with the following configuration:
```vim
let g:UltiSnipsExpandTrigger = '<C-s>'
```
But I cannot insert a `print` snippet with it.

Using the above `<c-u>` mapping, when I typed in 'print' and press <c-u>, it becomes:
```python
printncm2_ultisnips#expand_or("\", 'm')")
```

I don't understand. Any help is much appreciated.

**Comment by roxma** [comment]
@Neur1n 

My mistake, The `<c-u>` mapping should be written as `    imap <expr> <c-u>`.

**Comment by Neur1n** [comment]
@roxma 
Thanks a lot, it works now.

**Comment by jalcine** [comment]
Have any consideration gone into moving some of this pre-init logic into the plugin itself? Like under a `let g:ncm2#auto = 1` flag?

**Comment by roxma** [comment]
@jalcine

> Have any consideration gone into moving some of this pre-init logic into the plugin itself? Like under a let g:ncm2#auto = 1 flag?

No. My own configuration is different from this version. It's hard to generalize.

**Comment by uorbe001** [comment]
![screen shot 2018-08-24 at 12 02 57](https://user-images.githubusercontent.com/610728/44579130-b006c700-a795-11e8-8317-426af472ae66.png)

Does this parameter preview on the menu `(to, msg)` and the parenthesis completion depend on the source or a specific config option? I have it set up with lsp and I'm not getting that, so I wonder if it is because the source doesn't support that?

**Comment by filipekiss** [comment]
@uorbe001 from my experience with NCM, it depends on the source, not a config option.

**Comment by terkelg** [comment]
I'm not sure I completely understand what this line does:
```
inoremap <silent> <expr> <CR> ((pumvisible() && empty(v:completed_item)) ?  "\<c-y>\<cr>" : (!empty(v:completed_item) ? ncm2_ultisnips#expand_or("", 'n') : "\<CR>" ))
```

**Comment by roxma** [comment]
@terkelg

```
1. (pumvisible() && empty(v:completed_item)) -> 
    popup menu is visible but non was selected ->  "\<c-y>\<cr>" -> 
        close the popup with <c-y> and break line with <CR>

2. !empty(v:completed_item) -> 
    popup menu may or may not be visible, but there's a valid completion there -> 
        try expand its snippet, if expansion fails, do nothing else

3. empty(v:completed_item) -> 
    this implies that !pumvisible() in step 1, popup menu is not visible, no valid completion out there -> 
        break the line with <CR>
```

I don't remember why I chose this mapping. Currently I'm using `inoremap <silent> <expr> <CR> ncm2_ultisnips#expand_or("\<CR>", 'n')` in my vimrc.

**Comment by wizzup** [comment]
(Neovim)

Forgive me if it sound too stupid, I am very new to ncm.

Where is the declaration and what is the value of `$WORKSPACE_DIR`?

I am using [junegunn/vim-plug](https://github.com/junegunn/vim-plug) by manually install it

```bash
$ curl -fLo ~/.local/share/nvim/site/autoload/plug.vim --create-dirs \
    https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
```
and my current `init.vim` (`$MYVIMRC`) structure is like this

```vim
call plug#begin()
    Plug 'morhetz/gruvbox'
    Plug 'autozimu/LanguageClient-neovim', {'branch': 'next', 'do': './install.sh'}
call plug#end()

let g:LanguageClient_serverCommands = {
    \ 'haskell': ['hie-wrapper', '--lsp'],
    \ 'python': ['pyls'],
\ }
```

**Comment by roxma** [comment]
> Where is the declaration and what is the value of $WORKSPACE_DIR?

`$WORKSPACE_DIR`  is a environment variable points to my ncm2 development workspace. `:help let-$`

You could remove it since you  do `call plug#begin()` with no parameteres.

**Comment by ukrdev** [comment]
Sorry, but i don't understand, how to put function arguments on autocomplete.

**Comment by roxma** [comment]
@ukrdev you should be familiar with one of [ultisnips](https://github.com/SirVer/ultisnips), [neosnippet](https://github.com/Shougo/neosnippet.vim) or [snipmate](https://github.com/garbas/vim-snipmate) before using the snippet feature for parameter expansion.

**Comment by faulesocke** [comment]
I tried for half a day getting parameter expansion to work. I didn't even manage to make it work with the vimrc from above. When I type the name of the function, it shows the completion popup but when I select it and press Enter the dialog closes, a new line is entered but no parameter expansion happens. Am I missing a key or something?

I was trying with a rust source with LanguageClient-neovim and RLS nightly. Everything else works and I'm using that stuff for quite a while but parameter expansion just simply wont work. What am I missing?

**Comment by faulesocke** [comment]
Wow, it looks like I finally found it:

```vim
let g:LanguageClient_hasSnippetSupport = 1
```
